File: src/main/java/gargoyle/ct/pref/impl/prop/CTPrefProperty.java
package gargoyle.ct.pref.impl.prop;

import gargoyle.ct.convert.Converter;
import gargoyle.ct.convert.Converters;
import gargoyle.ct.log.Log;
import gargoyle.ct.pref.CTPreferencesProvider;
import gargoyle.ct.prop.impl.CTBaseObservableProperty;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Objects;
import java.util.prefs.BackingStoreException;
import java.util.prefs.Preferences;

public class CTPrefProperty<T> extends CTBaseObservableProperty<T> {
    @NotNull
    private final Converter<T> converter;
    @Nullable
    private final String def;
    private final Preferences preferences;

    protected CTPrefProperty(@NotNull Class<T> type, @NotNull CTPreferencesProvider provider, String name, T def) {
        this(type, Converters.get(type), provider, name, def);
    }

    protected CTPrefProperty(Class<T> type, @NotNull Converter<T> converter, CTPreferencesProvider provider, String name,
                             @Nullable T def) {
        super(type, name);
        this.converter = converter;
        this.def = def == null ? null : converter.format(def);
        preferences = provider.preferences();
    }

    @Override
    public final void set(@Nullable T value) {
        T oldValue = get();
        if (Objects.equals(oldValue, value)) {
            return;
        }
        preferences.put(name, value == null ? null : converter.format(value));//XXX
        sync();
        firePropertyChange(value, oldValue);
    }

    @Override
    public final T get() {
        return get(def == null ? null : converter.parse(def)); //XXX
    }

    @Override
    public T get(T def) {
        String value = preferences.get(name, null);
        if (value == null) {
            return def;
        }
        try {
            return converter.parse(value);
        } catch (IllegalArgumentException ex) {
            Log.warn(ex, ex.getMessage());
            return def;
        }
    }

    private void sync() {
        try {
            preferences.flush();
        } catch (BackingStoreException ex) {
            Log.error(ex, ex.getMessage());
        }
    }
}

File: src/main/kotlin/gargoyle/ct/pref/impl/prop/CTPrefProperty.kt
package gargoyle.ct.pref.impl.prop

import gargoyle.ct.convert.Converter
import gargoyle.ct.convert.Converters
import gargoyle.ct.log.Log
import gargoyle.ct.pref.CTPreferencesProvider
import gargoyle.ct.prop.impl.CTBaseObservableProperty
import java.util.prefs.BackingStoreException
import java.util.prefs.Preferences
import kotlin.reflect.KClass

open class CTPrefProperty<T : Any> protected constructor(
    type: KClass<T>, converter: Converter<T>, provider: CTPreferencesProvider, name: String,
    def: T?
) : CTBaseObservableProperty<T>(type, name) {
    private val converter: Converter<T>
    private val def: String?
    private val preferences: Preferences

    protected constructor(type: KClass<T>, provider: CTPreferencesProvider, name: String, def: T) : this(
        type,
        Converters.get<T>(type),
        provider,
        name,
        def
    )

    init {
        this.converter = converter
        this.def = if (def == null) null else converter.format(def)
        preferences = provider.preferences()
    }

    override fun set(value: T) {
        val oldValue = get()
        if (oldValue == value) {
            return
        }
        preferences.put(name, converter.format(value)) //XXX
        sync()
        firePropertyChange(value, oldValue)
    }

    override fun get(): T {
        return converter.parse(preferences[name, null] ?: def ?: error("")) //XXX
    }

    override fun get(def: T): T {
        val value = preferences[name, null] ?: return def
        return try {
            converter.parse(value)
        } catch (ex: IllegalArgumentException) {
            Log.warn(ex, ex.message ?: "")
            def
        }
    }

    private fun sync() {
        try {
            preferences.flush()
        } catch (ex: BackingStoreException) {
            Log.error(ex, ex.message ?: "")
        }
    }
}
